### ISSUE TITLE
[1.8] Meteor server blocked while rebuilding web.browser.legacy

### ISSUE THREAD

**Issue by sebakerckhof** [meteor/meteor#10261]
While Meteor 1.8 should improve rebuild times by delaying `web.browser.legacy`, in practice this doesn't seem to work for me.

Normally (e.g. in 1.7) the `Rebuild App` step is followed immediately with the `Server startup` step.
However, now, the process seems to hang after the `Rebuild App` step for the duration of the legacy step in which I can't reach my application. Only after the legacy build completes, the server starts, immediately followed by: `Finished delayed build of web.browser.legacy in 29752ms`. This of course renders the whole delayed building rather useless.

Also, the total rebuild time is now 20 seconds for `web` and 29 seconds for `web.legacy`.
While in Meteor 1.7 the total rebuild time for both combined was only 11 seconds. But I'll create a separate issue for that.

I've attached the cpu profile received from `qualia:profile`.
[profiles.zip](https://github.com/meteor/meteor/files/2464043/profiles.zip)

**Comment by benjamn** [comment]
The `web.browser.legacy` build is running in a separate process (the parent/build process) from the server startup code, so I believe those two chunks of work are racing in parallel, without blocking each other. If I hack the delayed build to take a long time (30 seconds) I can definitely see changes (or manually reload the app ) long before the legacy bundle finishes building. Of course there could be bugs that are affecting your experience and not mine, but I'm not sure what they could be, yet.

In any case, I think we should look into that performance regression first (the other issue)!

**Comment by benjamn** [comment]
One quick follow-up observation: if the legacy build is still running, and you make _another_ change, the build system won't pick that change up and start rebuilding until the previous legacy build is finished. So perhaps we should try to make the legacy builds cancelable somehow.

**Comment by KoenLav** [comment]
@benjamn for what it's worth; development speed on 1.8 is just terrible now.

Even though the build is delayed, we are still triggering two builds for every change. When a lot of small changes are made this significantly slows down the CPU. 

In my opinion there should definitely be an option to disable legacy building on development (or maybe it should even be the default).

**Comment by sebakerckhof** [comment]
@benjamn That's also what I would expect to happen, but here's a video of what is actually happening in my setup. Notice that around 00:39 the rebuild app finishes. Then I try to refresh the page, but the server is unresponsive for another 20 seconds until the legacy build finishes.

[meteor.mp4.zip](https://github.com/meteor/meteor/files/2466109/meteor.mp4.zip)

**Comment by benjamn** [comment]
@KoenLav Were you not using the 1.8 betas? I'm confused about what past you're referring to (compared to "now"), since I can't imagine that 1.8 is worse than 1.7.

I'm very much against disabling the legacy build by default. Please consider my previous idea of canceling the pending legacy build when another change takes place.

**Comment by KoenLav** [comment]
@benjamn we skipped 1.7 altogether, so I'm comparing with 1.6.

I understand why you might be against, because you want working apps for even the shittiest browser, but we are lucky to be gifted with an environment where we can expected all our users to be using modern browsers, so I couldn't care less.

Hence, for us, any performance loss caused by these legacy builds is unnecessary.

Right now (AFAIK) there isn't even an option to disable it, imposing extra CPU work on every Meteor development machine.

If the project needs to support legacy browsers (which is already a subset of projects) there are only two scenarios in which this work would be useful:
1. If some kind of compilation error would cause the legacy build to not be successful;
2. If some kind of compilation quirk would render different functional results between legacy and modern.

In short what we have to consider is; is it worth to impose the burden of double compilation on all developers, for the eventuality of some edge cases occurring in legacy compilation, whereas even when these edge cases occur they are only relevant for a subset of projects. 

In my opinion the answer to that question, by default, should be no. 

I can however imagine with different project experience you might answer yes to this question, but in my opinion at least the option to disable this should be provided.

Any other underlying performance issues should obviously also be looked into, but what were discussing now is more of an architectural consideration.

**Comment by sebakerckhof** [comment]
@KoenLav While I would also support a way to disable legacy altogether, Meteor's value proposition has always been to do everything with zero-config out-of-the-box. So as long as a large enough subset of projects needs to support older browsers, I understand the choice to have it enabled by default. But can we move this discussion to https://github.com/meteor/meteor-feature-requests ?

**Comment by KoenLav** [comment]
@sebakerckhof noted!

https://github.com/meteor/meteor-feature-requests/issues/333

The server should definitely not be blocked while this is happening :)

**Comment by benjamn** [comment]
> If I hack the delayed build to take a long time (30 seconds) I can definitely see changes (or manually reload the app) long before the legacy bundle finishes building.

I have to take this back. The way I was adding time was to include
```js
await new Promise(resolve => setTimeout(resolve, 15000));
```
[here](https://github.com/meteor/meteor/blob/2df1930bb08f7fae004f41373d3e2c23f8eec9b3/tools/isobuild/bundler.js#L3237), which is actually not a good simulation of `makeClientTarget` taking a long time, because nothing is happening in the parent process while that timeout is pending.

If I switch to the following synchronous busy-waiting `while` loop, then I see the symptoms you're describing:
```js
while (new Date - start < 15000) {}
```

What this suggests to me is that `makeClientTarget` is a very synchronous chunk of work without (m)any asynchronous pauses. This is good news, because it means we should be able to sprinkle some `async` functions and `await` expressions in that code to solve this problem.

**Comment by sebakerckhof** [comment]
@KoenLav note that legacy is a moving target (currently roughly browsers that support async/await). And because not all browsers are auto-updating, you may want to support legacy browsers some day in the future (e.g. a 2 year old safari browser). So in order not to break when the definition of legacy shifts due to some ES20xx features, it's better to have it enabled by default. It's easy enough to make an alias / npm script / whatever that inlcudes `--no-legacy`

**Comment by tafelito** [comment]
@benjamn in contrasts of what everyone is experiencing, I just updated to Meteor 1.8 and I'm still seeing that the legacy build happens right after the modern is built. I don't see it happening after the server started.

logs shows
```
=> Started proxy.
Building for web.browser
```

and right after that
```
Building for web.browser.legacy
```

Is this how it's supposed to work?

**Comment by benjamn** [comment]
@tafelito The first build works the same as always. The legacy build is delayed only when rebuilding after a file change.

**Comment by vjcspy** [comment]
any ideas help us to resolve this issue. I'm very confused why 1.8 take to much time to build web.browser.legacy. i upgraded 1.6 to 1.8 but in 1.6 i don't have any problems.

**Comment by sebakerckhof** [comment]
still relevant

**Comment by italojs** [comment]
We will remain closed until we have received the meteor 3.x report.

**Comment by zodern** [comment]
This issue still exists in Meteor 3. I fixed the main problem in Meteor 1.8.2, but there is still the issue of rebuilds having large amount of synchronous work that blocks the proxy in the Meteor tool.
